src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsKeyProcessorProvider.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            _globalOptionService = globalOptionService;
+        }
+
+        public KeyProcessor GetAssociatedProcessor(IWpfTextView wpfTextView)
+        {
+            return s_viewToProcessor.GetValue(wpfTextView, v => new InlineHintsKeyProcessor(_globalOptionService, v));
+        }
+
+        private class InlineHintsKeyProcessor : KeyProcessor
+        {
+            private readonly IGlobalOptionService _globalOptionService;
+            private readonly IWpfTextView _view;
+
+            public InlineHintsKeyProcessor(IGlobalOptionService globalOptionService, IWpfTextView view)
+            {
+                _globalOptionService = globalOptionService;
+                _view = view;
+                _view.Closed += OnViewClosed;
+                _view.LostAggregateFocus += OnLostFocus;
+            }
+
+            private static bool IsCtrlOrAlt(KeyEventArgs args)
+                => args.Key is Key.LeftCtrl or Key.RightCtrl or Key.LeftAlt or Key.RightAlt;
+
+            private Document? GetDocument()
+            {
+                var document =
+                    _view.BufferGraph.GetTextBuffers(b => true)
+                                     .Select(b => b.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges())
+                                     .WhereNotNull()
+                                     .FirstOrDefault();
+
+                return document;
+            }
+
+            private void OnViewClosed(object sender, EventArgs e)
+            {
+                // Disconnect our callbacks.
+                _view.Closed -= OnViewClosed;
+                _view.LostAggregateFocus -= OnLostFocus;
+
+                // Go back to off-mode just so we don't somehow get stuck in on-mode if the option was on when the view closed.
+                ToggleOff();
+            }
+
+            private void OnLostFocus(object sender, EventArgs e)
+            {
+                // if focus is lost (which can happen for shortcuts that include ctrl-alt...) then go back to normal
+                // inline-hint processing.
+                ToggleOff();
+            }
+
+            public override void KeyDown(KeyEventArgs args)
+            {
+                base.KeyDown(args);
+
+                // if this is either the ctrl or alt key, and only ctrl-alt is down, then toggle on. 
+                // otherwise toggle off if anything else is pressed down.
+                if (IsCtrlOrAlt(args))
+                {
+                    if (args.KeyboardDevice.Modifiers == (ModifierKeys.Control | ModifierKeys.Alt))
+                    {
+                        ToggleOn();
+                        return;
+                    }
+                }
+
+                ToggleOff();
+            }
+
+            public override void KeyUp(KeyEventArgs args)
+            {
+                base.KeyUp(args);
+
+                // If we've lifted a key up, then turn off the inline hints.
+                ToggleOff();
```

Comment:
Worth calling IsCtrlOrAlt() just so we're not rerunning the code in Toggle on _every_ keystroke?

Reply:
i can try that.  i'm being paranoid about something causing us to put this in the 'on' state and wanting to make sure we turn it off.